pyteal/ast/subroutine.py

```diff
@@ -188,14 +177,59 @@ def _arg_types_and_by_refs(
                 abi_output_kwarg[name] = expected_arg_type
                 continue
 
-            expected_arg_types.append(expected_arg_type)
-
+            arg_types.append(expected_arg_type)
             if expected_arg_type is ScratchVar:
                 by_ref_args.add(name)
             if isinstance(expected_arg_type, abi.TypeSpec):
                 abi_args[name] = expected_arg_type
 
-        return expected_arg_types, by_ref_args, abi_args, abi_output_kwarg
+        return (
+            impl_params,
+            annotations,
+            arg_types,
+            by_ref_args,
+            abi_args,
+            abi_output_kwarg,
+        )
+
+    @staticmethod
+    def _is_abi_annotation(obj: Any) -> bool:
+        try:
+            abi.type_spec_from_annotation(obj)
+            return True
+        except TypeError:
+            return False
+
+    @staticmethod
+    def _validate_annotation(
+        user_defined_annotations: dict[str, Any], parameter_name: str
+    ) -> type[Expr] | type[ScratchVar] | abi.TypeSpec:
+        ptype = user_defined_annotations.get(parameter_name, None)
+        if ptype is None:
+            # Without a type annotation, `SubroutineDefinition` presumes an implicit `Expr` declaration
+            # rather than these alternatives:
+            # * Throw error requiring type annotation.
+            # * Defer parameter type checks until arguments provided during invocation.
+            #
+            # * Rationale:
+            #   * Provide an upfront, best-effort type check before invocation.
+            #   * Preserve backwards compatibility with TEAL programs written
+            #     when `Expr` is the only supported annotation type.
+            # * `invoke` type checks provided arguments against parameter types to catch mismatches.
+            return Expr
+        elif ptype in (Expr, ScratchVar):
+            return ptype
+        elif SubroutineDefinition._is_abi_annotation(ptype):
+            return abi.type_spec_from_annotation(ptype)
+        else:
+            if not isclass(ptype):
+                raise TealInputError(
+                    f"Function has parameter {parameter_name} of declared type {ptype} which is not a class"
+                )
+            raise TealInputError(
+                f"Function has parameter {parameter_name} of disallowed type {ptype}. "
+                f"Only the types {(Expr, ScratchVar, 'ABI')} are allowed"
+            )
```

Comment:
```suggestion
        if ptype in (Expr, ScratchVar):
            return ptype
        if SubroutineDefinition._is_abi_annotation(ptype):
            return abi.type_spec_from_annotation(ptype)
        if not isclass(ptype):
            raise TealInputError(
                f"Function has parameter {parameter_name} of declared type {ptype} which is not a class"
            )
        raise TealInputError(
            f"Function has parameter {parameter_name} of disallowed type {ptype}. "
            f"Only the types {(Expr, ScratchVar, 'ABI')} are allowed"
        )
```

totally optional (equivalent since returning out of the else/if's)